File: src/main/java/com/xpdustry/nucleus/gatekeeper/GatekeeperDecision.java
// SPDX-License-Identifier: GPL-3.0-only
package com.xpdustry.nucleus.gatekeeper;

import java.time.Duration;

public sealed interface GatekeeperDecision permits GatekeeperDecision.Allow, GatekeeperDecision.Kick {

    Allow ALLOW = new Allow();

    final class Allow implements GatekeeperDecision {
        private Allow() {}
    }

    record Kick(String reason, Duration duration) implements GatekeeperDecision {

        private static final Duration FIVE_SECONDS = Duration.ofSeconds(5);

        public Kick(final String reason) {
            this(reason, FIVE_SECONDS);
        }
    }
}

File: src/main/kotlin/com/xpdustry/nucleus/gatekeeper/GatekeeperDecision.kt
// SPDX-License-Identifier: GPL-3.0-only
package com.xpdustry.nucleus.gatekeeper

import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds

sealed interface GatekeeperDecision {
    object Allow : GatekeeperDecision

    data class Kick(val reason: String, val duration: Duration = 5.seconds) : GatekeeperDecision
}
